mpqp/qasm/mpqp_to_qasm.py

```diff
@@ -0,0 +1,218 @@
+from mpqp.core.instruction.breakpoint import Breakpoint
+from mpqp.core.instruction.measurement import ExpectationMeasure
+from mpqp.core.instruction.gates.gate import SingleQubitGate
+from mpqp.core.instruction.gates import *
+from mpqp.qasm.open_qasm_2_and_3 import remove_user_gates
+
+
+def _handle_measurement(instruction: BasisMeasure) -> str:
+    if instruction.c_targets is None:
+        return "\n".join(
+            f"measure q[{target}] -> c[{i}];"
+            for i, target in enumerate(instruction.targets)
+        )
+    else:
+        return "\n".join(
+            f"measure q[{target}] -> c[{c_target}];"
+            for target, c_target in zip(instruction.targets, instruction.c_targets)
+        )
+
+
+def _handle_connection(instruction: Gate | Barrier) -> str:
+    control = ""
+    if isinstance(instruction, ControlledGate):
+        control = ",".join([f"q[{j}]" for j in instruction.controls]) + ","
+    return control + ",".join([f"q[{j}]" for j in instruction.targets])
+
+
+def _float_to_str(f: float) -> str:
+    if f.is_integer():
+        return str(int(f))
+    elif f % np.pi == 0:
+        return f"{int(f/np.pi)}*pi" if f != np.pi else "pi"
+    else:
+        return f"pi/{int(1 / f * np.pi)}" if (np.pi * (1 / f)).is_integer() else str(f)
+
+
+def _handle_rotation(gate: ParametrizedGate) -> str:
+    if gate.label in {"Rk", "CRk"}:
+        return "(" + _float_to_str(2 * np.pi / (2 ** float(gate.parameters[0]))) + ")"
+    return (
+        "(" + ",".join(_float_to_str(float(param)) for param in gate.parameters) + ")"
+    )
+
+
+def instruction_to_qasm(instruction: Instruction, simplify: bool = False) -> str:
+    if isinstance(instruction, CustomGate):
+        from qiskit.quantum_info.operators import Operator as QiskitOperator
+        from qiskit.qasm2.export import (
+            _instruction_call_site,  # pyright: ignore[reportPrivateUsage]
+            _define_custom_operation,  # pyright: ignore[reportPrivateUsage]
+        )
+        import collections.abc
+
+        gates_to_define: collections.OrderedDict[str, tuple[Instruction, str]] = (
+            collections.OrderedDict()
+        )
+
+        op = (
+            QiskitOperator(instruction.matrix)
+            .to_instruction()
+            ._qasm2_decomposition()  # pyright: ignore[reportPrivateUsage]
+        )
+        _define_custom_operation(op, gates_to_define)
+
+        gate_definitions_qasm = "\n".join(
+            f"{qasm}" for _, qasm in gates_to_define.values()
+        )
+        qasm_str = remove_user_gates(
+            "\n"
+            + gate_definitions_qasm
+            + "\n"
+            + _instruction_call_site(op)
+            + " "
+            + _handle_connection(instruction)
+            + ";"
+        )
+
+        return "\n" + qasm_str
+
+    elif isinstance(instruction, Gate):
+        instruction_str = (
+            instruction.qasm2_gate  # pyright: ignore[reportAttributeAccessIssue]
```

Comment:
this smells like architecture mishap, if we say that a gate "has to" have a
`qasm2_game`, then we wouldn't need this type ignore